### PR TITLE
BCDA-3722 - Update bcdaworker to leverage ServiceDate parameter

### PR DESCRIPTION
**Pull request by mtrang1263**
### Fixes [BCDA-3722](https://jira.cms.gov/browse/BCDA-3722)

Updates bcdaworker to use the serviceDate filter when querying from BFD.

### Change Details

1. Update bcdaworker and bb client to use the JobEnqueueArgs#ServiceDate parameter. When the field is set, we will query BFD with `service-date=le<SERVICE_DATE>`. See [RFC](https://github.com/CMSgov/beneficiary-fhir-data/blob/master/rfcs/0007-service-date-filter.md) for more details.
2. Consolidate bcda worker test setup into SetupSuite()/SetupTest()

### Security Implications

- [ ] new software dependencies
- [ ] security controls or supporting software altered
- [x] new data stored or transmitted

This PR will leverage service-date query parameter.

- [x] security checklist is completed for this change

[Security checklist](https://confluence.cms.gov/display/BCDA/20-10-16%3A+Service+Date+Filter+Implementation+Security+Checklist) for service-date filter implementation.

- [ ] requires more information or team discussion to evaluate security implications
- [x] no PHI/PII is affected by this change

### Acceptance Validation
1. CI Passes. Added unit tests to verify the service date parameter is passed in as expected.

We have a follow up [ticket](https://jira.cms.gov/browse/BCDA-3870) to implement smoke tests that will verify this behavior.

### CODE REVIEW

File: bcda/client/bluebutton.go [hunk 1]
```diff
@@ -133,8 +133,6 @@ func NewBlueButtonClient(config BlueButtonConfig) (*BlueButtonClient, error) {
 	return &BlueButtonClient{client, maxTries, retryInterval, config.BBServer, config.BBBasePath}, nil
 }
 
-type BeneDataFunc func(string, string, string, string, time.Time) (*models.Bundle, error)
```
Comment (mtrang1263) [inline bcda/client/bluebutton.go:136]: This type was remove because we no longer have the same function signature for all resource types (EOB has serviceDate parameter).

When we were using `BeneDataFunc` we introduced another interface (all BB funcs had the same signature) that had to be satisfied. Since serviceDate does not make sense w.r.t to Coverage and Patient types, it didn't make sense to update `BeneDataFunc` to include ServiceDate filter.

File: bcda/testUtils/client.go [hunk 2]
```diff
@@ -18,8 +18,8 @@ type BlueButtonClient struct {
 	MBI  *string
 }
 
-func (bbc *BlueButtonClient) GetExplanationOfBenefit(patientID, jobID, cmsID, since string, transactionTime time.Time) (*models.Bundle, error) {
-	args := bbc.Called(patientID)
+func (bbc *BlueButtonClient) GetExplanationOfBenefit(patientID, jobID, cmsID, since string, transactionTime, serviceDate time.Time) (*models.Bundle, error) {
```
Comment (mtrang1263) [inline bcda/testUtils/client.go:21]: This was changed so we can verify that all arguments are called as expected. Before, we just verified the patientID argument.

File: bcdaworker/main.go [hunk 3]
```diff
@@ -181,36 +181,35 @@ func createDir(path string) error {
...
-func writeBBDataToFile(ctx context.Context, bb client.APIClient, db *gorm.DB, acoID string, acoCMSID string, cclfBeneficiaryIDs []string, jobID, t, since string, transactionTime time.Time) (fileUUID string, error error) {
+func writeBBDataToFile(ctx context.Context, bb client.APIClient, db *gorm.DB, cmsID string, jobArgs models.JobEnqueueArgs) (fileUUID string, err error) {
 	segment := getSegment(ctx, "writeBBDataToFile")
 	defer func() {
 		if err := segment.End(); err != nil {
 			log.Error(err)
 		}
 	}()
 
-	if bb == nil {
```
Comment (mtrang1263) [inline bcdaworker/main.go:192]: Since writeBBDataToFile is package private, we can guarantee that client.APIClient is properly initialized. As such, we do not need a nil check.

File: bcdaworker/main.go [hunk 4]
```diff
@@ -181,36 +181,35 @@ func createDir(path string) error {
...
-		err := errors.New("Blue Button client is required")
-		log.Error(err)
-		return "", err
-	}
-
-	bbFunc := bbFuncByType(bb, t)
```
Comment (mtrang1263) [inline bcdaworker/main.go:198]: Replaced by the switch jobArgs.ResourceType

File: bcdaworker/main.go [hunk 5]
```diff
@@ -181,36 +181,35 @@ func createDir(path string) error {
...
-		err := fmt.Errorf("Invalid resource type requested: %s", t)
-		log.Error(err)
-		return "", err
-	}
-
-	if !utils.IsUUID(acoID) {
-		err := errors.New("Invalid ACO ID")
-		log.Error(err)
-		return "", err
+	var bundleFunc func(bbID string) (*fhirmodels.Bundle, error)
```
Comment (mtrang1263) [inline bcdaworker/main.go:192]: This only accepts bbID since all the other `Get...` methods accept arguments in the jobArgs and do not change between beneficiaries.

File: bcdaworker/main_test.go [hunk 6]
```diff
@@ -48,272 +50,255 @@ func (s *MainTestSuite) SetupSuite() {
...
+	s.db.Close()
+	os.RemoveAll(os.Getenv("FHIR_STAGING_DIR"))
+	os.RemoveAll(os.Getenv("FHIR_PAYLOAD_DIR"))
 }
 
 func TestMainTestSuite(t *testing.T) {
 	suite.Run(t, new(MainTestSuite))
 }
 
-func (s *MainTestSuite) TestWriteEOBDataToFile() {
```
Comment (mtrang1263) [inline bcdaworker/main_test.go:80]: Test was updated to request all resource types.

File: bcdaworker/main_test.go [hunk 7]
```diff
@@ -2,12 +2,12 @@ package main
 
 import (
```
Comment (mtrang1263) [inline bcdaworker/main_test.go:3]: Most of the changes to this file was a result of pushing up more state into `SetupSuite()` and `SetupTest()`

File: bcda/testUtils/utils.go [hunk 8]
```diff
@@ -20,17 +20,6 @@ func PrintSeparator() {
 	fmt.Println("**********************************************************************************")
 }
 
-func CreateStaging(jobID int) {
```
Comment (mtrang1263) [inline bcda/testUtils/utils.go]: Function was only used in main_test.go and is no longer needed.

File: bcda/testUtils/client.go (hunk 2 above)
Comment (PatKelsh) [reply bcda/testUtils/client.go:21]: good idea

**Review by PatKelsh — APPROVED** [review]

File: bcda/testUtils/utils.go [hunk 9]
```diff
@@ -20,21 +20,6 @@ func PrintSeparator() {
 	fmt.Println("**********************************************************************************")
 }
 
-func CreateStaging(jobID string) {
```
Comment (msnook) [inline bcda/testUtils/utils.go:23]: Why was this removed?

Comment (mtrang1263) [reply bcda/testUtils/utils.go:23]: Not sure why the comment fell off, but we only used this function in main_test.go. With the refactoring of the test setup, we no longer needed a dedicated function for this:

https://github.com/CMSgov/bcda-app/pull/576#discussion_r516309340

**Review by msnook — APPROVED** [review]
LGTM!
